homeassistant/components/cloud/helpers.py

```diff
@@ -0,0 +1,34 @@
+"""Helpers for the cloud component."""
+
+import logging
+import threading
+
+from homeassistant.core import HomeAssistant
+
+
+class FixedSizeQueueLogHandler(logging.Handler):
+    """Log handler to store messages, with auto rotation."""
+
+    MAX_RECORDS = 500
+
+    def __init__(self) -> None:
+        """Initialize a new LogHandler."""
+        super().__init__()
+        self._records: list[logging.LogRecord] = []
+        self._lock = threading.Lock()
+
+    def emit(self, record: logging.LogRecord) -> None:
+        """Store log message."""
+        with self._lock:
+            if len(self._records) >= self.MAX_RECORDS:
+                self._records.pop(0)
+            self._records.append(record)
+
+    async def get_logs(self, hass: HomeAssistant) -> list[str]:
+        """Get stored logs."""
+
+        def _get_logs() -> list[str]:
+            with self._lock:
```

Comment:
We try to avoid locks in the executor pool. Having an asyncio lock would be fine, but I guess that doesn't work with this logger interface.

Reply:
Maybe just copy the container before getting the log instead? Performance wise I think it's ok since getting the log is done infrequently.